[CORL-687] Webhooks

wyattjoh opened the pull request:
## What does this PR do?

Adds support for webhooks on the server for subscribing to events emitted by Coral.

## How do I test this PR?

TBD.

wyattjoh committed:
feat: initial webhook impl

wyattjoh committed:
feat: added support for key rotation

wyattjoh committed:
feat: harmonized fetcher

wyattjoh committed:
feat: added expired secrets cleaning

wyattjoh committed:
feat: event system refactor

wyattjoh committed:
feat: added story event

wyattjoh committed:
feat: simplfiied webhook handler

wyattjoh committed:
feat: added ref's to locations where user events can be added

wyattjoh committed:
feat: added UI to support webhooks

wyattjoh committed:
fix: renaming some Webhook -> WebhookEndpoint

wyattjoh committed:
fix: review comments to adjuist flow

wyattjoh committed:
feat: added localizations

wyattjoh committed:
fix: linting, updated snapshots

wyattjoh committed:
fix: adapted for new fluent

wyattjoh committed:
fix: rearranged folders

wyattjoh committed:
fix: linting

wyattjoh committed:
feat: added webhooks documentation

wyattjoh committed:
feat: improved toc generation

wyattjoh committed:
feat: added some tests to webhooks

alinhle committed:
fix: chain transition hooks

wyattjoh committed:
feat: added tests around webhook ui

wyattjoh committed:
fix: renamed events

wyattjoh committed:
fix: adjusted circle markdown linting

wyattjoh committed:
fix: adjusted doctoc script call

tessalt commented on WEBHOOKS.md:
````diff
@@ -0,0 +1,205 @@
+### How to verify the signature(s)
+
+```js
+// Set your secret secret here from the administratio n panel.
````
typo

tessalt commented on src/core/client/admin/routes/Configure/sections/WebhookEndpoints/ConfigureWebhookEndpoint/EndpointDangerZone.tsx:
```diff
@@ -0,0 +1,181 @@
+      </Localized>
+      <FormField>
+        <Localized id="configure-webhooks-rollSigningSecret">
+          <Label>Roll signing secret</Label>
```
isn't it usually "rotate"?

tessalt commented on src/core/client/admin/routes/Configure/sections/WebhookEndpoints/ConfigureWebhookEndpoint/RollSigningSecretModal.tsx:
```diff
@@ -0,0 +1,154 @@
+              <form onSubmit={handleSubmit}>
+                <HorizontalGutter size="double">
+                  <Localized id="configure-webhooks-rollSigningSecret">
+                    <Typography variant="header2">
```
I thought we weren't using `Typography` anymore post-redesign

tessalt commented on src/core/client/framework/lib/validation.tsx:
```diff
@@ -153,6 +154,15 @@ export const validateEqualPasswords = createValidator(
+ * validateEventSelection is a Validator that checks for a valid combination of
+ * event selections for webhook endpoints.
+ */
+export const validateEventSelection = createValidator(
```
Can we indicate that this is specific to webhooks in the name of this validator?

wyattjoh committed:
review: review fixes

alinhle commented on package.json:
```diff
@@ -388,7 +388,7 @@
   "husky": {
     "hooks": {
-      "pre-commit": "FAST_LINT=true lint-staged"
+      "pre-commit": "lint-staged"
```
The `FAST_LINT` is intentional because it's soooo much slower. It disables type information for linting. The idea was, that small changes, shouldn't make you wait so long for the linting when you commit. Obviously it makes sense when you change a lot of things, so it was a trade-off decision that favors convenience and speed over thorough linting, which still can be achieved by manually running `npm run lint`. Up to you :-)

alinhle commented on src/core/client/admin/routes/Configure/sections/WebhookEndpoints/ConfigureWebhookEndpoint/DeleteWebhookEndpointMutation.ts:
```diff
@@ -0,0 +1,38 @@
+      `,
+      variables: {
+        input: {
+          ...input,
```
It's usually better to write out the individual props instead of using `...` because it has been a constant source of bugs, e.g. `{ id: input.id }`.

alinhle commented:
![image](https://user-images.githubusercontent.com/14221600/73022177-56427a80-3e29-11ea-941b-95e4e9d3f207.png)
Can we have a similar solution to what @tessalt did with the multisite feature in order to avoid nested `<form>`?

alinhle commented:
One note from me: I feel like rotating the secret needs some kind of visually confirmation that it was successfully rotated.

alinhle commented on src/core/server/events/events.ts:
```diff
@@ -0,0 +1,87 @@
+
+export const StoryCreatedCoralEvent = createCoralEvent<
+  StoryCreatedCoralEventPayload
+>(CoralEventType.STORY_CREATED);
```
I think it should be possible to instead of:
```ts
export type CommentReleasedCoralEventPayload = CoralEventPayload<
  CoralEventType.COMMENT_RELEASED,
  CommentReleasedInput
>;

export const CommentReleasedCoralEvent = createCoralEvent<
  CommentReleasedCoralEventPayload
>(CoralEventType.COMMENT_RELEASED);
```
To do:
```ts
export const CommentReleasedCoralEvent = createCoralEvent<
  CommentReleasedInput
>(CoralEventType.COMMENT_RELEASED);
```
and have a Type function that returns the Payload whenever you need:
```ts
type Payload = CoralEventPayload<typeof CommentReleasedCoralEvent>;
```
But I haven't actually validated it yet.

wyattjoh replied:
I've looked into this too, but there's not really a clean way to extract the generic type there, and your proposal loses the binding of the data type to the type name, meaning we don't get the free type switching provided by the current system. It's not as pretty, but nothing wrong defining a type like that.

I was actually considering naming them the same:

```ts
export type CommentLeftModerationQueueCoralEventPayload = CoralEventPayload<
  CoralEventType.COMMENT_LEFT_MODERATION_QUEUE,
  CommentLeftModerationQueueInput
>;

export const CommentLeftModerationQueueCoralEvent = createCoralEvent<
  CommentLeftModerationQueueCoralEventPayload
>(CoralEventType.COMMENT_LEFT_MODERATION_QUEUE);
```

becomes:

```ts
export type CommentLeftModerationQueueCoralEvent = CoralEventPayload<
  CoralEventType.COMMENT_LEFT_MODERATION_QUEUE,
  CommentLeftModerationQueueInput
>;

export const CommentLeftModerationQueueCoralEvent = createCoralEvent<
  CommentLeftModerationQueueCoralEvent
>(CoralEventType.COMMENT_LEFT_MODERATION_QUEUE);
```

And Typescript would select the type/value depending on the usage automatically.

wyattjoh committed:
review: review comments

wyattjoh committed:
review: adjusted signing secret confirmation

wyattjoh committed:
review: adjusted styles to harmonize button usage

wyattjoh committed:
fix: updated snapshots and tests

wyattjoh committed:
review: move form out of webhooks

Moved the form out of the webhooks by relocating the layout used for the
route associated with the configure routes.

wyattjoh committed:
fix: fixed bugs and snapshots with tests

wyattjoh committed:
feat: revised slack message format to use block api

wyattjoh committed:
fix: fixed a small text bug

kgardnr committed:
Merge branch 'master' into feature/CORL-687
